File: 3312-sorted-gcd-pair-queries/3312-sorted-gcd-pair-queries.py
class Solution(object):
    def gcdValues(self, nums, queries):
        """
        :type nums: List[int]
        :type queries: List[int]
        :rtype: List[int]
        """
        maximum = 0

        for num in nums:
            if num > maximum:
                maximum = num

        frequency = [0] * (maximum + 1)

        for num in nums:
            frequency[num] += 1

        exact_gcd = [0] * (maximum + 1)

        for g in range(maximum, 0, -1):
            divisible_count = 0

            multiple = g
            while multiple <= maximum:
                divisible_count += frequency[multiple]
                multiple += g

            
            pair_count = divisible_count * (divisible_count - 1) // 2

            
            multiple = g * 2
            while multiple <= maximum:
                pair_count -= exact_gcd[multiple]
                multiple += g

            exact_gcd[g] = pair_count

        
        prefix_pairs = [0] * (maximum + 1)

        for g in range(1, maximum + 1):
            prefix_pairs[g] = prefix_pairs[g - 1] + exact_gcd[g]

        answer = []

        for query in queries:
            
            left = 1
            right = maximum

            while left < right:
                middle = (left + right) // 2

                if prefix_pairs[middle] > query:
                    right = middle
                else:
                    left = middle + 1

            answer.append(left)

        return answer

File: 3312-sorted-gcd-pair-queries/3312-sorted-gcd-pair-queries.java
class Solution {
    public int[] gcdValues(int[] nums, long[] queries) {
        int maximum = 0;

        for (int num : nums) {
            if (num > maximum) {
                maximum = num;
            }
        }

        long[] frequency = new long[maximum + 1];

        for (int num : nums) {
            frequency[num]++;
        }

        
        long[] exactGcd = new long[maximum + 1];

        for (int g = maximum; g >= 1; g--) {
            long divisibleCount = 0;

            
            for (int multiple = g;
                 multiple <= maximum;
                 multiple += g) {
                divisibleCount += frequency[multiple];
            }

            
            long pairCount =
                divisibleCount * (divisibleCount - 1) / 2;

            
            for (int multiple = g * 2;
                 multiple <= maximum;
                 multiple += g) {
                pairCount -= exactGcd[multiple];
            }

            exactGcd[g] = pairCount;
        }

        
        long[] prefixPairs = new long[maximum + 1];

        for (int g = 1; g <= maximum; g++) {
            prefixPairs[g] =
                prefixPairs[g - 1] + exactGcd[g];
        }

        int[] answer = new int[queries.length];

        for (int i = 0; i < queries.length; i++) {
            long query = queries[i];

            int left = 1;
            int right = maximum;

            while (left < right) {
                int middle = left + (right - left) / 2;

                if (prefixPairs[middle] > query) {
                    right = middle;
                } else {
                    left = middle + 1;
                }
            }

            answer[i] = left;
        }

        return answer;
    }
}